
## Benchmarking Micrite

We created 'dilution series' for testing by combining human + microbial reads in different proportions.
Reads were generated by ART paired-end read simulation from reference genomes listed below. 
Each test FASTQ contains 1000 reads in total. 

We test micrite on hundreds of simulated datasets described in [micritebench](https://github.com/selkamand/micritebench)

Note we benchmark across multiple human reference genomes because many Kraken2 databases will not contain 
sufficient diversity in the human DNA included. We want to have some indication on how micrite performs on diverse populations.

human1 = [chm13v2.0_noY.fa.gz](https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/analysis_set/chm13v2.0_noY.fa.gz) (telomere to telomere assembly)
human2 = [GCF_000001405.40](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.40/) (GRCh38)
human3 = [GCA_011064465.2](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_011064465.2/) (Ashkenazi Human Reference Genome)
human4 = [GCA_018873775.2](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_018873775.2/) (Puerto-Rican Genome)
human5 = [GCA_003610035.2](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_003610035.2/) (Pangenome representing genomic regions from African-Ancestry Populations in the Americas that are not present in GRCh38)

ebv = [GCF_002402265.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002402265.1/) 
hpv16 = [GCF_000863945.3](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000863945.3/)
hpv18 = [GCF_000865665.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000865665.1/)

hpylori = [GCF_000307795.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000307795.1/)



We include a urls.txt file with download links for all genomes we used interest.



test_human_ebv.fa = mix 90 reads of human DNA + 10 reads from EBV
